RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc

```diff
@@ -68,25 +70,29 @@ void HGCalImagingAlgo::populate(const HGCRecHitCollection& hits){
 // with different input (reset should be called between events)
 void HGCalImagingAlgo::makeClusters()
 {
-
-  // used for speedy search
-  std::vector<KDTree> hit_kdtree(2*(maxlayer+1));
-
+  std::vector<std::vector<std::vector< KDNode> > >  layerClustersPerLayer(2*(maxlayer+2));
+  edm::SerialTaskQueue layersQueue;
   //assign all hits in each layer to a cluster core or halo
-  for (unsigned int i = 0; i <= 2*maxlayer+1; ++i) {
-    KDTreeBox bounds(minpos[i][0],maxpos[i][0],
-		     minpos[i][1],maxpos[i][1]);
-
-    hit_kdtree[i].build(points[i],bounds);
-
-    unsigned int actualLayer = i > maxlayer ? (i-(maxlayer+1)) : i; // maps back from index used for KD trees to actual layer
-
-    double maxdensity = calculateLocalDensity(points[i],hit_kdtree[i], actualLayer); // also stores rho (energy density) for each point (node)
-    // calculate distance to nearest point with higher density storing distance (delta) and point's index
-    calculateDistanceToHigher(points[i]);
-    findAndAssignClusters(points[i],hit_kdtree[i],maxdensity,bounds,actualLayer);
-  }
-  //make the cluster vector
+  tbb::parallel_for(size_t(0), size_t(2*maxlayer+2), [&](size_t i) {
```

Comment:
@Dr15Jones 
what will this do in the current framework context?

I suppose in a one-thread job this will stay executing on a single thread.
In a multi-thread nThread, this algo is running as a part of a stream module. Will it get to steal tasks from the remaining nThread-1 threads?

I'm thinking if we should already enable a similar feature in RecoTracker/CkfPattern/src/CkfTrackCandidateMakerBase.cc as well
@VinInn

Reply:
>I suppose in a one-thread job this will stay executing on a single thread.

Correct
>In a multi-thread nThread, this algo is running as a part of a stream module. Will it get to steal tasks from the remaining nThread-1 threads?

It is allowed to do so. Using `tbb::this_task_arena::isolate` should prevent that.

Reply:
>>    In a multi-thread nThread, this algo is running as a part of a stream module. Will it get to steal tasks from the remaining nThread-1 threads?
> It is allowed to do so. Using tbb::this_task_arena::isolate should prevent that.

I see that you proposed to use it in https://github.com/cms-sw/cmssw/pull/22620#discussion_r174597347 . 
So, when ```tbb::this_task_arena::isolate``` is in place, wouldn't this just become a regular sequential loop in the stream module call context?

Reply:
>So, when tbb::this_task_arena::isolate is in place, wouldn't this just become a regular sequential loop in the stream module call context?

No. Other TBB threads which are not presently running a framework module would be able to run tasks generated by `parallel_for`.

Reply:
Actually I have already done this test:
here you see (sort by CPU time) the parallel:
http://fpantale.web.cern.ch/fpantale/parallel_layerCluster_PU200.html

and the original:
http://fpantale.web.cern.ch/fpantale/original_layer_cluster_PU200.html

when running with 24 threads and 1 stream.
They produce the same results.

Reply:
timereport with isolate
http://fpantale.web.cern.ch/fpantale/parallel_isolate_layerCluster_PU200.html